File: core-examples/src/main/java/io/vertx/example/core/http/proxy/Proxy.java
package io.vertx.example.core.http.proxy;

import io.vertx.core.Future;
import io.vertx.core.VerticleBase;
import io.vertx.core.http.HttpClient;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.launcher.application.VertxApplication;

/*
 * @author <a href="http://tfox.org">Tim Fox</a>
 */
public class Proxy extends VerticleBase {

  public static void main(String[] args) {
    VertxApplication.main(new String[]{Proxy.class.getName()});
  }

  @Override
  public Future<?> start() throws Exception {
    HttpClient client = vertx.createHttpClient();
    return vertx.createHttpServer().requestHandler(serverRequest -> {
      System.out.println("Proxying request: " + serverRequest.uri());
      serverRequest.pause();
      HttpServerResponse serverResponse = serverRequest.response();
      client.request(serverRequest.method(), 8282, "localhost", serverRequest.uri())
        .onSuccess(clientRequest -> {
          clientRequest.headers().setAll(serverRequest.headers());
          clientRequest.send(serverRequest).onSuccess(clientResponse -> {
            System.out.println("Proxying response: " + clientResponse.statusCode());
            serverResponse.setStatusCode(clientResponse.statusCode());
            serverResponse.headers().setAll(clientResponse.headers());
            serverResponse.send(clientResponse);
          }).onFailure(err -> {
            System.out.println("Back end failure");
            serverResponse.setStatusCode(500).end();
          });
        }).onFailure(err -> {
          System.out.println("Could not connect to localhost");
          serverResponse.setStatusCode(500).end();
        });
    }).listen(8080);
  }
}

File: core-examples/src/main/kotlin/io/vertx/example/core/http/proxy/Proxy.kt
package io.vertx.example.core.http.proxy

import io.vertx.core.http.HttpClientOptions
import io.vertx.kotlin.core.http.*

class Proxy : io.vertx.core.AbstractVerticle()  {
  override fun start() {
    var client = vertx.createHttpClient(HttpClientOptions())
    vertx.createHttpServer().requestHandler({ serverRequest ->
      println("Proxying request: ${serverRequest.uri()}")
      serverRequest.pause()
      var serverResponse = serverRequest.response()
      client.request(serverRequest.method(), 8282, "localhost", serverRequest.uri()).onSuccess({ clientRequest ->
        clientRequest.headers().setAll(serverRequest.headers())
        clientRequest.send(serverRequest).onSuccess({ clientResponse ->
          println("Proxying response: ${clientResponse.statusCode()}")
          serverResponse.setStatusCode(clientResponse.statusCode())
          serverResponse.headers().setAll(clientResponse.headers())
          serverResponse.send(clientResponse)
        }).onFailure({ err ->
          println("Back end failure")
          serverResponse.setStatusCode(500).end()
        })
      }).onFailure({ err ->
        println("Could not connect to localhost")
        serverResponse.setStatusCode(500).end()
      })
    }).listen(8080)
  }
}
